src/expr/src/scalar/like_pattern.rs

```diff
@@ -73,3 +147,745 @@ pub fn build_regex(pattern: &str, flags: &str) -> Result<Regex, EvalError> {
+// followed by (non-wildcard) literal characters. The last sub-pattern may
+// have an empty SUFFIX.
+//
+// Example: the PATTERN "n__dl%" can be broken into the following parts:
+//   1. SUB-PATTERN = <WILDCARDS ""> <SUFFIX "n">
+//   2. SUB-PATTERN = <WILDCARDS "__"> <SUFFIX "dl">
+//   3. SUB-PATTERN = <WILDCARDS "%"> <SUFFIX "">
+//
+// The WILDCARDS can be any combination of '_', which matches exactly 1 char,
+// and '%' which matches zero or more chars. These wildcards can be simplified
+// down to the (min, max) of characters they might consume:
+//     ""  = (0, 0)    // doesn't consume any characters
+//     "_" = (1, 1)    // consumes exactly one
+//     "%" = (0, many) // zero or more
+// These are additive, so:
+//     "_%"    = (1, many)
+//     "__%__" = (4, many)
+//     "%%%_"  = (1, many)
+
+// Matches a chain of SUB-PATTERNs
+#[derive(Debug, Clone, Deserialize, Serialize)]
+struct StringMatcher {
+    /// The minimum number of characters that can be consumed by the wildcard expression.
+    consume: usize,
+    /// Whether the wildcard expression can consume an arbitrary number of characters.
+    many: bool,
+    /// A string literal that is expected after the wildcards.
+    suffix: String,
+    /// Any sub-patterns that come after this one.
+    next: Option<Box<StringMatcher>>,
+}
+
+impl StringMatcher {
+    pub fn is_match(&self, text: &str) -> bool {
+        let mut rest = text;
+        // Go ahead and skip the minimum number of characters the sub-pattern consumes:
+        if self.consume > 0 {
+            let mut chars = rest.chars();
+            for _ in 0..self.consume {
+                if chars.next().is_none() {
+                    return false;
+                }
+            }
+            rest = chars.as_str();
+        }
+        if self.many {
+            // The sub-pattern might consume any number of characters, but we need to find
+            // where it terminates so we can match any subsequent sub-patterns. We do this
+            // by searching for the suffix string using str::find.
+            //
+            // We could investigate using a fancier substring search like Boyer-Moore:
+            // https://en.wikipedia.org/wiki/Boyer%E2%80%93Moore_string-search_algorithm
+            //
+            // .. but it's likely not worth it. It's slower for small strings,
+            // and doesn't really start outperforming the naive approach until
+            // haystack sizes of 1KB or greater. See benchmarking results from:
+            // https://github.com/killerswan/boyer-moore-search/blob/master/README.md
+            //
+            // Another approach that may be interesteing to look at is a
+            // hardware-optimized search:
+            // http://0x80.pl/articles/simd-strfind.html
+            if self.suffix.len() == 0 {
+                // Nothing to find... This should only happen in the last sub-pattern.
+                // TODO: Should this assert that self.next.is_none()?
```

Comment:
Any and all assertions are much welcome given that we rely on fuzzing/randomized testing for testing the SQL engine.